### PR TITLE
Deployment Config in YAML

### PR DESCRIPTION
**Pull request by jspreddy**
## Proposed changes

- Be able to provide the deployment config as a yaml file.  https://github.com/databrickslabs/dbx/issues/39
- Makefile updates to help with developer experience.
    - `pyenv` required by makefile.
- Contributor doc updates
- Changelog updates

## Types of changes

What types of changes does your code introduce to dbx?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Ready for review. Please let me know if any other tests are needed.

### CODE REVIEW

**Comment by jspreddy** [comment]
ready for review ^

**Comment by jspreddy** [comment]
rebased onto the latest changes.

**Comment by jspreddy** [comment]
make_file_self_documentation

File: dbx/utils/common.py [hunk 1]
```diff
@@ -69,17 +76,56 @@ def get_context() -> Any:
             return None
 
 
-class DeploymentFile:
+class AbstractDeploymentConfig(ABC):
```
Comment (jspreddy) [inline dbx/utils/common.py:79]: This ABC `AbstractDeploymentConfig` and the concrete classes `JsonDeploymentConfig` and `YamlDeploymentConfig` is the core of the logic changes.

File: dbx/utils/common.py [hunk 2]
```diff
@@ -69,17 +76,56 @@ def get_context() -> Any:
...
+
+class JsonDeploymentConfig(AbstractDeploymentConfig):
+    def get_environment(self, environment: str) -> Any:
+        return read_json(self._path).get(environment)
+
+    def get_all_environment_names(self) -> Any:
         return list(read_json(self._path).keys())
 
 
+def get_deployment_config(path: str) -> AbstractDeploymentConfig:
```
Comment (jspreddy) [inline dbx/utils/common.py:119]: This function returns the proper config class.

File: dbx/utils/common.py [hunk 3]
```diff
@@ -143,12 +189,6 @@ def create_context(self, payload):
         return result
 
 
-def dbx_echo(message: str):
```
Comment (jspreddy) [inline dbx/utils/common.py:146]: moved this function to top of file.

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,12 @@
 [tool.black]
 line-length = 120
-target-version = ['py37']
\ No newline at end of file
+target-version = ['py37']
+
+
+[tool.pytest.ini_options]
```
Comment (jspreddy) [inline pyproject.toml:6]: pytest can read config from `pyproject.toml`

File: tests/unit/test_adjust_path.py [hunk 5]
```diff
@@ -1,16 +1,27 @@
...
+
+
+def format_path(rel_path: str):
+    """Format proper path for a given relative path or a glob path"""
+    # required for when pytest is NOT ran from within the tests/unit dir.
+    return os.path.join(os.path.dirname(__file__), rel_path)
 
 
 class AdjustPathTest(unittest.TestCase):
     def test_single_node(self):
```
Comment (jspreddy) [inline tests/unit/test_adjust_path.py:18]: This test was a false positive before. Fixed it.

File: tests/unit/test_adjust_path.py [hunk 6]
```diff
@@ -22,12 +33,48 @@ def test_single_node(self):
...
                 package_requirement,
                 _file_uploader,
+                api_client,
             )
 
             for job_spec in deployment.get("jobs"):
                 for key, value in job_spec.items():
                     self.assertIsNotNone(value)
 
+    def test_that_upstream_files_are_checked_and_paths_are_properly_re_written(self):
```
Comment (jspreddy) [inline tests/unit/test_adjust_path.py:43]: Added another test to check the files are uploaded and paths are properly re-written

File: tests/unit/test_common.py [hunk 7]
```diff
@@ -1,6 +1,36 @@
...
+import os
+
+
+def format_path(rel_path: str):
+    """Format proper path for a given relative path or a glob path"""
+    # required for when pytest is NOT ran from within the tests/unit dir.
+    return os.path.join(os.path.dirname(__file__), rel_path)
+
+
+class CommonUnitTest(unittest.TestCase):
```
Comment (jspreddy) [inline tests/unit/test_common.py:13]: Added tests to validate that the config reader classes are working as expected.

**Comment by renardeinside** [comment]
hi @jspreddy , 

the change itself looks good to me, could you please add some docs & examples to the quickstart file [in this section](https://github.com/databrickslabs/dbx/blob/master/docs/source/quickstart.rst#preparing-deployment-file)?

**Comment by jspreddy** [comment]
- Added targets to help with generating docs.
- Added documentation for yaml config files.
- Fixed a minor mistake in the doc for json schema.

**Comment by renardeinside** [comment]
hi @jspreddy , 
this is a great job, I'll merge this and will start preparing the 0.1.5 release.

**Comment by jspreddy** [comment]
When do you plan on releasing?

**Comment by renardeinside** [comment]
hi @jspreddy , today in the late evening approx. need to do some preps

**Pull request merged**